text/3476-expose-fn-type.md

```diff
@@ -0,0 +1,164 @@
+    inner: goods
+};
+```
+Here, we make a `FnContainer` which can hold every function with the signature `() -> ()` via generics.
+But what about explicitly designing the `FnContainer` for a specific function, just like the compiler does when resolving the generics. This will work the same as with the trait impl from above:
+```rust
+struct GoodsContainer {
+    inner: fn goods,
+}
+fn goods() {}
+
+let contained_goods = GoodsContainer {
+    inner: goods,
+}
+```
+---
+A function with a more complex signature, like with parameters, modifiers or a return type, is still just referenced by its name, because it's already unique
+```rust
+async fn request_name(id: PersonID) -> String { .. }
+
+impl Requestable for fn request_name {
+    /* ... */
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+As described in the **Guide-level explanation**, with the syntax `fn <fn_path>`, we can reference the type behind the named function.
+
+When the function is for example in a different mod, it should be referenced by its path
+```rust
+mod sub {
+    fn sub_mod_fn() { .. }
+}
+impl fn sub::sub_mod_fn {
+    /* ... */
+}
+```
+
+It should be also possible to get the type of functions inside impl blocks:
+
+```rust
+struct MyStruct;
+impl MyStruct {
+    fn new() -> Self { Self }
+}
+impl fn MyStruct::new {
+    /* ... */
+}
+```
+
+When a function has generics, they will be handled as follows, just like we know it from normal types
+```rust
+fn send<T: Send>(val: T) {}
+impl<T: Send> ParcelStation for fn send<T> {
+    /* ... */
+}
+```
+
+When we have an implicit generic, they will be appended in order to the generic list:
+```rust
+fn implicit_generic(val: impl Clone) -> impl ToString {}
+impl<T: Send, U: ToString> for fn implicit_generic<T, U> {
+    /* ... */
+}
+```
+
+Just as structs and enums have the possibility to derive traits to automatically generate code, function type do too
+
+```rust
+#[derive(DbgSignature)]
+fn signature_test(val: i32) -> bool {
+    /* ... */
+}
+
+// Expands to
+
+fn signature_test(val: i32) -> bool {
+    /* ... */
+}
+impl DbgSignature for fn signature_test {
+    fn dbg_signature() -> &'static str {
+        "fn signature_test(val: i32) -> bool"
+    }
+}
+```
+
+Other than that, it should behave like every other type does.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- When introducing the derive feature, it could lead to parsing problems with proc macros having an older `syn` crate version.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The type behind functions already exists, we just need to expose it to the user.
+The hard part would be allowing derives, because that may break some things.
+
+# Prior art
+[prior-art]: #prior-art
+
+i dont know any
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Is the syntax good? It could create confusion between a function pointer.
```

Comment:
Not familiar with compiler parsing internals, but could also consider `fn {path}`, which is slightly similar to the current compiler diagnostics.

Reply:
isn't that what i already do? or do you mean with explicit `{}` wrapped around?

Reply:
yes